crates/metaslang/bindings/src/lib.rs

```diff
@@ -67,37 +85,49 @@ impl<KT: KindTypes + 'static> Bindings<KT> {
             tree_cursor,
         );
         builder
-            .build(file_path, &builder::NoCancellation, |handle, cursor| {
-                self.cursors.insert(handle, cursor.clone());
-            })
+            .build(&builder::NoCancellation)
             .expect("Internal error while building bindings");
+        self.cursors.extend(builder.extract_cursors());
+        self.definiens.extend(builder.extract_definiens());
         builder
     }
 
-    pub fn all_definitions(&self) -> impl Iterator<Item = Handle<'_, KT>> + '_ {
+    pub fn all_definitions(&self) -> impl Iterator<Item = Definition<'_, KT>> + '_ {
         self.stack_graph
             .iter_nodes()
             .filter(|handle| self.stack_graph[*handle].is_definition())
-            .map(|handle| Handle {
+            .map(|handle| Definition {
                 owner: self,
                 handle,
             })
     }
 
-    pub fn all_references(&self) -> impl Iterator<Item = Handle<'_, KT>> + '_ {
+    pub fn all_references(&self) -> impl Iterator<Item = Reference<'_, KT>> + '_ {
         self.stack_graph
             .iter_nodes()
             .filter(|handle| self.stack_graph[*handle].is_reference())
-            .map(|handle| Handle {
+            .map(|handle| Reference {
                 owner: self,
                 handle,
             })
     }
 
-    pub fn cursor_to_handle(&self, cursor: &Cursor<KT>) -> Option<Handle<'_, KT>> {
+    pub fn definition_at(&self, cursor: &Cursor<KT>) -> Option<Definition<'_, KT>> {
         for (handle, handle_cursor) in &self.cursors {
-            if handle_cursor == cursor {
-                return Some(Handle {
+            if handle_cursor == cursor && self.stack_graph[*handle].is_definition() {
```

Comment:
I see we are using the default equality implementation to compare cursors. Would that be enough?

- Two cursors can point to the same node but have different fields, for example `is_complete`, or even `parent`.
- Two cursors (in two different files with identical source code) might have the same values, even if they don't point to the same node.

I think we might need to brainstorm on cursor/node identity here.

Reply:
This is a placeholder implementation and very inefficient (ie. it's O(n) with the number of graph nodes). We need a way to either attach arbitrary data to cursors, or make them hashable/orderable so that we can use them as keys in some map.

> * Two cursors can point to the same node but have different fields, for example `is_complete`, or even `parent`.

Is this possible? I understand the `is_complete` field, but `parent` too? How?

> * Two cursors (in two different files with identical source code) might have the same values, even if they don't point to the same node.

Hmm... I would have thought equality for `Rc` would compare pointers, but it seems I'm mistaken. So, then definitely no, default cursor equality is not enough.

---

I now realize that CST nodes *do* have an `.id()` which we can use, which is a `usize`. It's already used by the graph builder to index cursors as syntax nodes (although btw it's [being downcasted to u32](https://github.com/manastech/slang/blob/feat%2Fimport-resolver/crates/metaslang/graph_builder/src/graph.rs#L55) :thinking:). I'll change the implementation.

Reply:
:heavy_check_mark:

Reply:
> Is this possible? I understand the is_complete field, but parent too? How?

if you use `cursor.spawn()` it will clone the current cursor, but with no ancestors. This is used to constrain the clone to the sub-tree only.

Reply:
QQ: with this implementation, if a file is deleted/removed, and re-added, do we retain any cursors or edges to it from/to other files? or would this API to get defs/refs just recalculate everything using the new cursor/node ID from the new file?

Reply:
> if you use cursor.spawn() it will clone the current cursor, but with no ancestors. This is used to constrain the clone to the sub-tree only.

Oh! I see. I hadn't thought of that.

> QQ: with this implementation, if a file is deleted/removed, and re-added, do we retain any cursors or edges to it from/to other files? or would this API to get defs/refs just recalculate everything using the new cursor/node ID from the new file?

There is no way to remove nodes from the stack graph. So any invalidation of the original source files implies we need to rebuild from scratch. If you added a file multiple times, it would build up on the number of refs/defs in that file and also retain the cursors. The `jump_to_definition` function will find new paths from a given reference every time, so if you added a file containing the definition multiple times, the algorithm will find several possible paths. Currently it will only return the longest only though, and only one path.

Reply:
How about the other way around?
if a file contains references, and the user changed its contents a few times, would the graph retain the old nodes to the old references?
that means we would return `Reference` handles to nodes that the user already discard/overwrote ...

Not blocking for this PR, but I wonder how easy/cheap is it to handle this use case, and actually remove nodes for an old file when it is deleted?
If it is too much work, or means a complete rewrite of the stack graph crate, maybe we can just prevent deleting files (for now at least).